### PR TITLE
MM/POSIX: check if there is enough space for the posix backing file.

### PR DESCRIPTION
**Pull request by alinask**
solves #392

### CODE REVIEW

**Comment by alinask** [comment]
in this commit, we still can't specify the path to use in case shm_open isn't used.
this one checks that we have enough memory to use in /dev/shm and if not, a proper message is printed out to the user. (instead of it failing on a bus error)

**Comment by miked-mellanox** [comment]
not all systems have /dev/shm, what happens if /dev/shm does not exist?

**Comment by alinask** [comment]
next commit will take care of this. if shm_open won't succeed, open() will be used with a path from the user.

File: src/uct/sm/mm/mm_posix.c [hunk 1]
```diff
@@ -31,11 +31,60 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
   {NULL}
 };
 
+static ucs_status_t uct_posix_test_mem(size_t length, int shm_fd)
+{
+    int *buf;
+    int i, iters, chunk_size = 256 * 1024;
+    size_t size_to_write, remaining, single_write, total_write = 0;
+
+    buf = (int*) ucs_malloc(length, "write buffer");
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: no need for explicit cast to int* here

File: src/uct/sm/mm/mm_posix.c [hunk 2]
```diff
@@ -31,11 +31,60 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+    memset(buf, 0, length);
+
+    if (length > chunk_size) {
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: it can be written simpler - just loop while remaining>0 and keep doing remaining-=size_to_write

File: src/uct/sm/mm/mm_posix.c [hunk 3]
```diff
@@ -31,11 +31,60 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
 };
 
+static ucs_status_t uct_posix_test_mem(size_t length, int shm_fd)
+{
+    int *buf;
+    int i, iters, chunk_size = 256 * 1024;
+    size_t size_to_write, remaining, single_write, total_write = 0;
+
+    buf = (int*) ucs_malloc(length, "write buffer");
+    memset(buf, 0, length);
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: shouldn't the size of buf be chunk_size?

File: src/uct/sm/mm/mm_posix.c [hunk 4]
```diff
@@ -31,11 +31,60 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+        if (single_write <= 0) {
+            ucs_error("Failed to write remaining %zu bytes. %m", remaining);
+        } else {
+            total_write += single_write;
+        }
+    }
+
+    ucs_free(buf);
+
+    if (length != total_write) {
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: you can check for every write that it returned the expected value, and fail immediately.

File: src/uct/sm/mm/mm_posix.c [hunk 5]
```diff
@@ -31,11 +31,60 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+        size_to_write = chunk_size;
+        remaining = length % chunk_size;
+    } else {
+        iters = 1;
+        size_to_write = length;
+        remaining = 0;
+    }
+
+    for (i = 0; i < iters; i++) {
+        single_write = write(shm_fd, buf, size_to_write);
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: need to lseek to 0 in the beginning

File: src/uct/sm/mm/mm_posix.c [hunk 6]
```diff
@@ -31,11 +31,60 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+            total_write += single_write;
+        }
+    }
+
+    ucs_free(buf);
+
+    if (length != total_write) {
+        ucs_error("Failed to write to the backing file. "
+                  "Not enough memory to write %zu bytes. "
+                  "Please check that /dev/shm or the directory you specified has "
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: need to pass /dev/shm as a parameter

File: src/uct/sm/mm/mm_posix.c [hunk 7]
```diff
@@ -31,11 +31,60 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+            ucs_error("Failed to write remaining %zu bytes. %m", remaining);
+        } else {
+            total_write += single_write;
+        }
+    }
+
+    ucs_free(buf);
+
+    if (length != total_write) {
+        ucs_error("Failed to write to the backing file. "
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: The error can say just "not enough space in %s"

File: src/uct/sm/mm/mm_posix.c (hunk 6 above)
Comment (alinask) [inline src/uct/sm/mm/mm_posix.c]: after i add the choice of path, it wont necesseraly be /dev/shm and then passing the path to this function is either /dev/shm or the fill path to user's path.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1069/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/497/ for details (Mellanox internal link).

File: src/uct/sm/mm/mm_posix.c (hunk 3 above)
Comment (alinask) [inline src/uct/sm/mm/mm_posix.c]: why..? the copy is from it to the file refered to by shm_fd, in chunk_size blocks.

Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: because you always copy in chunk_size blocks, no need for a larger buffer.

File: src/uct/sm/mm/mm_posix.c (hunk 6 above)
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: ok

**Comment by shamisp** [comment]
autobot build

File: src/uct/sm/mm/mm_posix.c (hunk 2 above)
Comment (alinask) [inline src/uct/sm/mm/mm_posix.c]: i could get to a negative value then for remaining. if i still have to take care of the leftover, isn't it the same?
or you meant something else?

Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: size_to_write=ucs_min(remaining, chunk_size)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1078/ for details.

File: src/uct/sm/mm/mm_posix.c [hunk 8]
```diff
@@ -31,11 +31,59 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+    buf = ucs_malloc(chunk_size, "write buffer");
+    if (buf == NULL) {
+        ucs_error("Failed to allocate memory for testing space for backing file.");
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    memset(buf, 0, chunk_size);
+    if (lseek(shm_fd, 0, SEEK_SET) < 0) {
+        status = UCS_ERR_IO_ERROR;
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: pls add error print

File: src/uct/sm/mm/mm_posix.c [hunk 9]
```diff
@@ -31,11 +31,59 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+        goto out_free_buf;
+    }
+
+    remaining = length;
+    while (remaining > 0) {
+        size_to_write = ucs_min(remaining, chunk_size);
+        single_write = write(shm_fd, buf, size_to_write);
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: the type of single_write should be ssize_t

File: src/uct/sm/mm/mm_posix.c [hunk 10]
```diff
@@ -31,11 +31,59 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+        if (single_write < size_to_write) {
+            ucs_error("Failed to write %zu bytes. %m", size_to_write);
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: are you sure errno would be set in case single_write is >=0 by < size_to_write? according to man page, "On error, -1 is returned, and errno is set appropriately".

File: src/uct/sm/mm/mm_posix.c [hunk 11]
```diff
@@ -31,11 +31,59 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+            goto out_check;
+        } else {
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: no need for 'else' since you have 'goto'

File: src/uct/sm/mm/mm_posix.c [hunk 12]
```diff
@@ -31,11 +31,59 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+            total_write += single_write;
+        }
+        remaining -= size_to_write;
+    }
+
+out_check:
+    if (length != total_write) {
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: why do you need to check this if you check already that every single write is successful?

File: src/uct/sm/mm/mm_posix.c [hunk 13]
```diff
@@ -31,11 +31,59 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+
+out:
+    return status;
+}
+
 static ucs_status_t
 uct_posix_alloc(uct_pd_h pd, size_t *length_p, ucs_ternary_value_t hugetlb,
                 void **address_p, uct_mm_id_t *mmid_p UCS_MEMTRACK_ARG)
 {
-    ucs_status_t status = UCS_ERR_NO_MEMORY;
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: why is that needed?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/509/ for details (Mellanox internal link).

File: src/uct/sm/mm/mm_posix.c (hunk 10 above)
Comment (alinask) [inline src/uct/sm/mm/mm_posix.c]: right, it doesnt. this is from before i changed the 'if'. before the CR, it only checked if single_write is <0 and thats why errno was printed. 
will remove the errno print.

File: src/uct/sm/mm/mm_posix.c (hunk 12 above)
Comment (alinask) [inline src/uct/sm/mm/mm_posix.c]: in case 'write' was successful, making sure in the end that all of 'length' was written.

Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: is there any way all writes were successful and the length is not the same?

Comment (alinask) [inline src/uct/sm/mm/mm_posix.c]: i can change:
if (single_write < size_to_write) 
to 
if (single_write != size_to_write) 
and then the answer is no...

Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: yeah, it can't be > anyway.

File: src/uct/sm/mm/mm_posix.c [hunk 14]
```diff
@@ -31,11 +31,59 @@ static ucs_config_field_t uct_posix_pd_config_table[] = {
...
+    if (lseek(shm_fd, 0, SEEK_SET) < 0) {
+        status = UCS_ERR_IO_ERROR;
+        goto out_free_buf;
+    }
+
+    remaining = length;
+    while (remaining > 0) {
+        size_to_write = ucs_min(remaining, chunk_size);
+        single_write = write(shm_fd, buf, size_to_write);
+        if (single_write < size_to_write) {
```
Comment (yosefe) [inline src/uct/sm/mm/mm_posix.c]: need to check that single_write < 0 in separate 'if' which prints %m (errno).
otherwise, it's >=0. and then we can compare to size_to_write

Comment (alinask) [inline src/uct/sm/mm/mm_posix.c]: this is what i had before, the separate 'if's..
will change.

**Comment by yosefe** [comment]
:+1:

**Comment by shamisp** [comment]
:+1: 

Let's wait for mellanox test results to merge it in.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1080/ for details.

**Comment by alinask** [comment]
18:38:54 [ RUN      ] ud/test_uct_perf.envelope/1
18:38:54 [jenkins01:7001 :0] timer_wheel.c:58   Fatal: Timer resolution is too low. Min resolution 49710.269630 usec, wanted 48083.565185 usec

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/511/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
Shell we ignore the failure ?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1087/ for details.

**Comment by alinask** [comment]
@yosefe can this be merged?

**Comment by yosefe** [comment]
:+1:
